src/value.jl

```diff
@@ -1,145 +1,194 @@
-function CategoricalValue(level::Integer, pool::CategoricalPool{T, R}) where {T, R}
-    return CategoricalValue(convert(R, level), pool)
+# categorical value concept
+iscatvalue(::Type) = false
+iscatvalue(x::Any) = iscatvalue(typeof(x))
+
+# categorical value concept implementation for CategoricalValue and CategoricalString
+iscatvalue(::Type{<:CategoricalString}) = true
+iscatvalue(::Type{<:CategoricalValue}) = true
+valtype(::Type{<:CategoricalValue{T}}) where T = T
+valtype(::Type{<:CategoricalString}) = String
+# type of levels
+reftype(::Type{<:CategoricalValue{T, R}}) where {T,R} = R
+reftype(::Type{<:CategoricalString{R}}) where R = R
+
+pool(x::CatValue) = x.pool
+level(x::CatValue) = x.level
+
+# extract the type of original value from categorical value type
+unwrap_catvalue_type(::Type{T}) where {T<:CatValue} = valtype(T)
+unwrap_catvalue_type(::Type{Union{T, Null}}) where {T<:CatValue} = Union{valtype(T), Null}
+unwrap_catvalue_type(::Type{T}) where {T} = T
+
+# default categorical value type for given non-null value type `T` and reference type `R`
+catvalue_type(::Type{T}, refType::Type{R}) where {T<:CatValue, R} =
```

Comment:
The intention is to be agnostic about the nature of `T`.
We don't know what `T` is, but if it's using the same `refType`, we keep using it.
If `T` uses a different reftype, the new categorical value type is constructed.

Reply:
`T<:CatValue`, so we completely know what it is. And even if we didn't, it would be quite weird to preserve `T` when `reftype(T) === R`, but not preserve it in other cases. `R` is really a secondary parameter compared to `valtype(T)`, it shouldn't be allowed to determine the kind of `CatValue` you get.

Reply:
The idea is to try to keep "value type" `T`, so that `convert()` doesn't do unnecessary copies. The `valtype(T)` would be preserved anyway.

Reply:
I don't see why `convert` would make a copy since `catvalue_type` should return the same type as `T` in the end, right?

Reply:
Yes. Right now there's only `CategoricalString` and `CategoricalValue` that wrap distinct types (`AbstractString` and everything else), so the default `catvalue_type(valtype(T), R)` would match `T`.
But I hope all the types are resolved at compile time, so `reftype(T) === R` check, which might be relevant for the future, costs nothing.

Reply:
I agree it costs nothing in terms of computation, but it costs in terms of complexity, and I'm not convinced it's useful right now. Let's design the simplest system which works now, and only make it more general later if we feel the need for it. It doesn't make much sense to design a general system without using it -- anyway, if we don't test its general capabilities, there will always be an undetected failure somewhere.